[CI/CD] lint, test, release chart

paologallinaharbur opened the pull request:
- [x] Moving e2e charts to /.charts
 - [x] adapt pipelines
 - [x] Test implementation in fork

kang-makes commented on .github/workflows/release-integration.yml:
```diff
@@ -91,18 +77,21 @@ jobs:
+      - name: Push pre-release images
+        if: ${{ github.event.release.prerelease }}
         run: |
+          echo "DOCKER_IMAGE_TAG=${DOCKER_IMAGE_TAG}-pre" >> $GITHUB_ENV
```
I know that you have tested in a fork but are we sure that every time you "touch" github's envfile the environment is instantaneously updated?

I had the feeling that is not happening and that is the reason I [used the environment variable directly.](https://github.com/newrelic/newrelic-k8s-metrics-adapter/pull/40/files#diff-2c3b5a007a3ca5391253871b413bf6d78192fe551b28bdfe115df67aa07528b1R78) and it is working:
 - From prod repo: https://hub.docker.com/layers/newrelic-k8s-metrics-adapter/newrelic/newrelic-k8s-metrics-adapter/0.3.0-pre/images/sha256-3436a0523cf0394b9e1f373976fb80436aa3f053d8b42eab4e9a39bac3411b2d?context=explore
 - From fork repo: https://hub.docker.com/r/jperezflorez123/newrelic-k8s-metrics-adapter/tags

paologallinaharbur replied:
I think it does, a race condition like that would be horrible, but you are right, it is better to use classical Env Var since we are in the same step!

kang-makes commented on .gitignore:
```diff
@@ -5,5 +5,5 @@ bin
 
-e2e/charts/e2e-resources/Chart.lock
-e2e/charts/e2e-resources/charts/
+charts/e2e-resources/Chart.lock
+charts/e2e-resources/charts/
```
It would be awesome that this change had its own commit. Reviewing this would be easier ;)

paologallinaharbur replied:
![image](https://user-images.githubusercontent.com/43335750/159679585-2aeaaae9-67fa-4c2e-baa3-0c851288b477.png)

kang-makes commented on e2e/charts/e2e-resources/Chart.yaml:
```diff
@@ -1,10 +0,0 @@
-apiVersion: v2
-version: 1.2.0
-appVersion: "1.0.0"
-description: This chart creates e2e resources.
-name: e2e-resources
-
-dependencies:
-  - name: kube-state-metrics
-    version: 2.13.2
-    repository: https://kubernetes.github.io/kube-state-metrics
```
Why is this deleted? Is this not going to make this chart uninstallable?

kang-makes replied:
Ah. It has so many changes that git thinks is a new file: https://github.com/newrelic/nri-kubernetes/pull/377/files#diff-a8dbaa85f24d99fa37816cd13be16708b3eab7f9b9fd58fb2908d1ff073f5b6aR1-R24

Ok.

kang-makes commented on charts/e2e-resources/Chart.yaml:
```diff
@@ -0,0 +1,24 @@
+apiVersion: v2
+version: 1.2.0
+appVersion: "1.0.0"
```
In case we release it by mistake. `appVersion` is not mandatory.
```suggestion
version: 1.2.0-devel
```

kang-makes commented on .github/workflows/release-chart.yml:
```diff
@@ -0,0 +1,48 @@
+        with:
+          fetch-depth: 0
+
+      # As of today, there is no way to filter/exclude charts to be released. Config does not support it:
+      # https://github.com/helm/chart-releaser/blob/4eb598f96aa53d1c0b8234af2087834e1f2275e1/pkg/config/config.go#L40-L61
+      # So we delete the Chart.yaml to filter the chart: https://github.com/helm/chart-releaser-action/pull/18
+      - name: remove the chart template as it should not be deployed
+        run: |
+          rm ./charts/e2e-resources/Chart.yaml
```
Would we want to release this with a `-devel` or something? It would be useful to have something like this in the canaries someday?

I would leave this as you have it right now until we need it, just asking myself.

paologallinaharbur replied:
Yes we could do it, but since it is not currently needed I will keep it like that.

kang-makes commented on .github/workflows/release-integration.yml:
```diff
@@ -39,46 +39,32 @@ jobs:
-      - if: ${{ github.event.release.prerelease }}
-        run: |
-          echo "DOCKER_IMAGE_TAG=${DOCKER_IMAGE_TAG}-pre" >> $GITHUB_ENV
-          echo "S3_PATH=$S3_PATH/test" >> $GITHUB_ENV
```
Docker `-pre` suffix has been moved to a step below. What about this `S3_PATH`?

paologallinaharbur replied:
S3_PATH is not useful anymore, since we are not publishing things in S3

kang-makes reviewed (commented):
I left a few comments and doubts here and there. It is a pretty good job overall.

paologallinaharbur commented:
I rebased and moved around things, but the changes with respect to the old pipeline are just:
```git
index 872250d..c2a4585 100644
--- a/.github/workflows/release-integration.yml
+++ b/.github/workflows/release-integration.yml
@@ -84,7 +84,7 @@ jobs:
       - name: Push pre-release images
         if: ${{ github.event.release.prerelease }}
         run: |
-          echo "DOCKER_IMAGE_TAG=${DOCKER_IMAGE_TAG}-pre" >> $GITHUB_ENV
+          DOCKER_IMAGE_TAG=${DOCKER_IMAGE_TAG}-pre"
           docker buildx build --push --platform=$DOCKER_PLATFORMS \
             -t $DOCKER_IMAGE_NAME:$DOCKER_IMAGE_TAG \
             .
diff --git a/charts/e2e-resources/Chart.yaml b/charts/e2e-resources/Chart.yaml
index ef96b38..49e879c 100755
--- a/charts/e2e-resources/Chart.yaml
+++ b/charts/e2e-resources/Chart.yaml
@@ -1,6 +1,5 @@
 apiVersion: v2
-version: 1.2.0
-appVersion: "1.0.0"
+version: 1.2.0-devel
 description: This chart creates e2e resources for nri-kubernetes.
 name: e2e-resources
```

kang-makes commented on .github/workflows/push_pr.yaml:
```diff
@@ -31,9 +32,43 @@ jobs:
+          if [[ -n "$changed" ]]; then
+            echo "::set-output name=changed::true"
+          fi
+      - name: Set up helm-unittest
+        if: steps.list-changed.outputs.changed == 'true'
+        run: |
+          helm plugin install https://github.com/quintush/helm-unittest
+      - name: Run unit tests
+        if: steps.list-changed.outputs.changed == 'true'
+        run: |
```
Could we do it in the same step?

Like in the "original" one: https://github.com/newrelic/newrelic-k8s-metrics-adapter/pull/40/files#diff-944291df2c9c06359d37cc8833d182d705c9e8c3108e7cfe132d61a06e9133ddR78-R87

paologallinaharbur replied:
sure!

kang-makes commented on .github/workflows/release-chart.yml:
```diff
@@ -0,0 +1,48 @@
+jobs:
+#   Sometimes chart-releaser might fetch an outdated index.yaml from gh-pages, causing a WAW hazard on the repo
+#   This job checks the remote file is up to date with the local one on release
+  validate-gh-pages-index:
+    runs-on: ubuntu-latest
+    steps:
+      - name: Checkout
+        uses: actions/checkout@v2
+        with:
+          ref: gh-pages
+      - name: Download remote index file and check equality
+        run: |
+          curl -vsSL https://newrelic.github.io/nri-kubernetes/index.yaml > index.yaml.remote
+          LOCAL="$(md5sum < index.yaml)"
+          REMOTE="$(md5sum < index.yaml.remote)"
+          echo "$LOCAL" = "$REMOTE"
+          test "$LOCAL" = "$REMOTE"
```
I'd like to pint that this step can fail if there is no `index.yaml` created in the `gh-pages` branch.

But I swa you put [an index.yaml in that branch](https://github.com/newrelic/nri-kubernetes/commit/1ae507d0c2572740219614886f53ca28948c3289). Are we completely sure that this is the bare minimum for `cr` to work? Because I put a completely empty `index.yaml` (`touch index.yaml && git add -A && git push`) and it failed.

I would change my PR on the metrics adapter if you confirm me that this works :)

paologallinaharbur replied:
I run 
```
git checkout --orphan gh-pages
 - Deleted everything but the gitignore
helm repo index .
```

In the forked it worked!

kang-makes replied:
Really cool. I did not think about using `helm` to create an empty index. Now I feel dumb 😅

kang-makes reviewed (approved):
A couple of nits but LGTM.

Separating the commit about changing the e2e chart location made the re-review way easier. Good job!

roobre commented on .github/workflows/release-chart.yml:
```diff
@@ -0,0 +1,48 @@
+        with:
+          fetch-depth: 0
+
+      # As of today, there is no way to filter/exclude charts to be released. Config does not support it:
+      # https://github.com/helm/chart-releaser/blob/4eb598f96aa53d1c0b8234af2087834e1f2275e1/pkg/config/config.go#L40-L61
+      # So we delete the Chart.yaml to filter the chart: https://github.com/helm/chart-releaser-action/pull/18
```
Alternatively, could we move the e2e chart to a different dir? Like `charts-internal` or something along those lines.

I'm trying to avoid footgunning ourselves if we need to do a release manually and we forget that we need to remove this (which we will).

paologallinaharbur commented on .github/workflows/release-integration.yml:
```diff
@@ -91,18 +77,21 @@ jobs:
         if: ${{ ! github.event.release.prerelease }}
         run: |
           docker buildx build --push --platform=$DOCKER_PLATFORMS \
+            -t $DOCKER_IMAGE_NAME:$DOCKER_IMAGE_TAG \
             -t $DOCKER_IMAGE_NAME:latest \
```
this is @kang-makes black magic

kang-makes replied:
These are tags. You can [get variables](https://github.com/txqueuelen/pdns-stateless/blob/main/.github/workflows/periodic-upgrade.yaml#L30-L33) and use it [as many times as you need](https://github.com/txqueuelen/pdns-stateless/blob/main/.github/workflows/periodic-upgrade.yaml#L53-L58)

paologallinaharbur replied:
yeah, I actually think it is quite standard to use them 😅  I was not sure if Roberto was kidding or not, thus I accused you of black magic 😄

paologallinaharbur commented on .github/workflows/push_pr.yaml:
```diff
@@ -31,9 +32,41 @@ jobs:
   test:
     name: Run tests
-    needs: [ build ]
+    needs: [ build, chart-lint ]
```
not at all

paologallinaharbur commented on .github/workflows/release-chart.yml:
```diff
@@ -0,0 +1,48 @@
+          ref: gh-pages
+      - name: Download remote index file and check equality
+        run: |
+          curl -vsSL https://newrelic.github.io/nri-kubernetes/index.yaml > index.yaml.remote
```
I would not change it since every-other pipeline has it hardcoded. Moreover, it is impossible not to notice it in case you mistype it since the pipeline would fail

paologallinaharbur commented:
waiting for the ci to kick in, to merge this
